zerver/lib/markdown/__init__.py

```diff
@@ -2083,15 +2083,22 @@ def make_md_engine(realm_filters_key: int, email_gateway: bool) -> None:
 # function on the URLs; they are expected to be HTML-escaped when
 # rendered by clients (just as links rendered into message bodies
 # are validated and escaped inside `url_to_a`).
-def topic_links(realm_filters_key: int, topic_name: str) -> List[str]:
-    matches: List[str] = []
+def topic_links(realm_filters_key: int, topic_name: str) -> List[Dict[str, Union[str, int]]]:
```

Comment:
This type doesn't need a union; we're stripping away the `index` field before returning.

Reply:
Indeed, however, mypy shouts at me with 
```
(zulip-py3-venv) vagrant@23d7671b6078:/srv/zulip$ ./tools/run-mypy
zerver/lib/markdown/__init__.py:2121: error: Incompatible return value type 
(got "List[Dict[str, Union[str, int]]]", expected "List[Dict[str, str]]")  [return-value]
Found 1 error in 1 file (checked 1184 source files)
```
Wondering if it Is because I am returning back the same `matches` object? Any ideas?

Reply:
Yeah, that makes sense. Maybe copying to a new List would help here. I'll run some experiments. Maybe this is something they've addressed already but it does look like a hard problem to track which elements of each dict have been deleted with just static type checks so I doubt there's a solution without cloning the List.

Reply:
Yeah, mypy doesn't like changing the type of a variable.  Generally code is more readable using a new variable name for the output (vs. the input), anyway, so I'd recommend doing that.

Reply:
Hmmm... this is proving to be tricky. Seems like even using a new variable and copying to it the k, v pair seems to unsettle mypy. 
```
    .....
    matches = sorted(matches, key=lambda k: k["index"])
    matches_output: List[Dict[str, str]] = [{k: v for k, v in match.items() if k != "index"} for match in matches]
    return matches_output
```
still throws 
```
zerver/lib/markdown/__init__.py:2118: error: Value expression in dictionary comprehension has incompatible type "Union[str, int]"; expected type "str"  [misc]
Found 1 error in 1 file (checked 1184 source files)
```

Hmmm...Seems like only type-casting the value in the dictionary before returning seems to work.
```
    return [{k: str(v) for k, v in match.items() if k != "index"} for match in matches]
```

Reply:
Hmm, yeah, I think that's reasonable.